subprojects/tooling-api/src/crossVersionTest/groovy/org/gradle/integtests/tooling/r56/TestLauncherDebugTestsCrossVersionTest.groovy

```diff
@@ -0,0 +1,152 @@
+import com.sun.jdi.connect.Connector
+import com.sun.jdi.connect.ListeningConnector
+import com.sun.jdi.connect.TransportTimeoutException
+import org.gradle.integtests.tooling.fixture.TargetGradleVersion
+import org.gradle.integtests.tooling.fixture.ToolingApiSpecification
+import org.gradle.integtests.tooling.fixture.ToolingApiVersion
+import org.gradle.tooling.BuildException
+import spock.lang.Timeout
+
+
+@ToolingApiVersion(">=5.6")
+@TargetGradleVersion(">=5.6")
+@Timeout(60)
+class TestLauncherDebugTestsCrossVersionTest extends ToolingApiSpecification {
+
+    def setup() {
+        buildFile << """
+            plugins { id 'java-library' }
+            ${mavenCentralRepository()}
+            dependencies { testCompile 'junit:junit:4.12' }
+        """
+        file('src/test/java/example/MyTest.java').text = """
+            package example;
+            public class MyTest {
+                @org.junit.Test public void foo() throws Exception {
+                     org.junit.Assert.assertEquals(1, 1);
+                }
+            }
+        """
+        file('src/test/java/example/SecondTest.java').text = """
+            package example;
+            public class SecondTest {
+                @org.junit.Test public void bar() throws Exception {
+                     org.junit.Assert.assertEquals(1, 1);
+                }
+            }
+        """
+    }
+
+    def "build fails if debugger is not ready"() {
+        when:
+        withConnection { connection ->
+            connection.newTestLauncher()
+                .withJvmTestClasses("example.MyTest")
+                .debugTests(findFreePort())
+                .run()
+        }
+
+        then:
+        thrown(BuildException)
+    }
+
+    def "can launch tests in debug mode"() {
+        setup:
+        def (connector, port, arguments) = startSocketListenDebugSession()
+
+        when:
+        withConnection { connection ->
+            connection.newTestLauncher()
+                .withJvmTestClasses("example.MyTest")
+                .debugTests(port)
+                .run()
+        }
+
+        then:
+        true // test successfully executed with debugger attached
+
+        cleanup:
+        connector.stopListening(arguments)
+    }
+
+    def "Forks only one JVM to debug"() {
+        setup:
+        buildFile << """
+             tasks.withType(Test) {
+                  forkEvery = 1
+                  maxParallelForks = 2
+            }
+        """
+        def (connector, port, arguments) = startSocketListenDebugSession()
+
+        when:
+        withConnection { connection ->
+            connection.newTestLauncher()
+                .withJvmTestClasses("example.MyTest", "example.SecondTest")
+                .debugTests(port)
+                .run()
+        }
+
+        then:
+        true // test successfully executed with debugger attached
+
+        cleanup:
+        connector.stopListening(arguments)
+    }
+
+    int findFreePort() {
```

Comment:
I think we have a utility class that already does this somewhere... take a look at PortAllocator

Reply:
I made use of it, thanks for the suggestion.